mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocFinder.java

```diff
@@ -95,53 +94,59 @@ public long find(CRUDOperationContext ctx,
                 cursor = cursor.sort(mongoSort);
                 LOGGER.debug("Result set sorted");
             }
-            int size = cursor.count();
-            long ret = size;
-            LOGGER.debug("Applying limits: {} - {}", from, to);
-
-            if (from != null) {
-                cursor.skip(from.intValue());
-                if (to != null && from > to) //if 'to' is not null but lesser than 'from', skip the entire list to return nothing 
-                {
-                    cursor.skip(size);
+            int numMatched = cursor.count();
+            int nRetrieve=numMatched;
+            List<DBObject> mongoResults=null;            
+            List<JsonDoc> jsonDocs=null;
+
+           LOGGER.debug("Applying limits: {} - {}", from, to);
+            long retrievalTime=0;
+
+            // f and t are from and to indexes, both inclusive
+            int f=from==null?0:from.intValue();
+            int t=to==null?numMatched-1:to.intValue();
+            if(f<0)
+                f=0;
+            if(t>=numMatched)
+                t=numMatched-1;
+            if(t<f||f>=numMatched)
```

Comment:
I don't think it matters much, but I think if f >= numMatched, then t will be < f, since t is at most numMatched - 1, so `||f>=numMatched`) is maybe not necessary.

IMHO, you could consider skipping this check, and only do `int nRetrieve = Math.min(0, t - f + 1)`. Then you only need to check if nRetrieve is > 0, and it doesn't need to be reassigned again on 116.

Many ways to skin a cat, so all I really mean to say is that I think you can take out `||f>=numMatched` :-)

Other than that, LGTM